apps/emqx/src/emqx_persistent_message_ds_replayer.erl

```diff
@@ -268,23 +270,26 @@ replay_range(
         end,
     MessagesReplay = [emqx_message:set_flag(dup, true, Msg) || Msg <- MessagesUnacked],
     %% Asserting that range is consistent with the message storage state.
-    {Replies, Until} = publish(FirstUnacked, MessagesReplay),
+    {Replies, Until} = publish(FirstUnacked, MessagesReplay, false),
     %% Again, we need to keep the iterator pointing past the end of the
     %% range, so that we can pick up where we left off.
     Range = Range0#ds_pubrange{iterator = ItNext},
     {Range, Replies ++ Acc};
 replay_range(Range0 = #ds_pubrange{type = checkpoint}, _AckedUntil, Acc) ->
     {Range0, Acc}.
 
-publish(FirstSeqno, Messages) ->
-    lists:mapfoldl(
-        fun(Message, Seqno) ->
-            PacketId = seqno_to_packet_id(Seqno),
-            {{PacketId, Message}, next_seqno(Seqno)}
-        end,
-        FirstSeqno,
-        Messages
-    ).
+publish(FirstSeqNo, Messages, PreserveQos0) ->
+    do_publish(FirstSeqNo, Messages, PreserveQos0, []).
+
+do_publish(SeqNo, [], _, Acc) ->
+    {lists:reverse(Acc), SeqNo};
+do_publish(SeqNo, [#message{qos = 0} | Messages], false, Acc) ->
+    do_publish(SeqNo, Messages, false, Acc);
+do_publish(SeqNo, [#message{qos = 0} = Message | Messages], true, Acc) ->
+    do_publish(SeqNo, Messages, true, [{undefined, Message} | Acc]);
```

Comment:
```suggestion
do_publish(SeqNo, [#message{qos = 0} | Messages], false = PreserveQoS0, Acc) ->
    do_publish(SeqNo, Messages, PreserveQoS0, Acc);
do_publish(SeqNo, [#message{qos = 0} = Message | Messages], true = PreserveQoS0, Acc) ->
    do_publish(SeqNo, Messages, PreserveQoS0, [{undefined, Message} | Acc]);
```

Reply:
I'm not sure if this change adds to readability. It's unlikely that the reader will forget that the 3rd argument stands for `PreserveQoS0` while reading a function that is less than 10 lines long.

Reply:
Another option is to use `preserve_qos0` and `dont_preserve_qos0` atoms to overcome desemantisation 😅

Reply:
Nit:
Maybe it would be useful to highlight that `undefined` here is used as a _valid_ `packet_id` ... because we should go through many modules to find how `undefined` finally gets ignored.
Btw this is not even highlighted by the `emqx_message:to_packet/2` type

```erlang
-spec to_packet(emqx_types:packet_id(), emqx_types:message()) ->
    emqx_types:packet().
```

Reply:
Thanks, I fixed the signature for that function.